{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "588549f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob,os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "%matplotlib widget\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import json\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from mpl_toolkits.mplot3d import Axes3D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "562df732",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_expression_data(direction):\n",
    "    \"\"\"\n",
    "    This function read in all .JSON files along the given direction. \n",
    "    Involved conditions, expression data under each condition,\n",
    "    replicates under each condition, and the set of genes will be extracted. \n",
    "    They will be returned as lists.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read the expression .json files based on the direction path. \n",
    "    os.chdir(direction)\n",
    "    files = []\n",
    "    expr_files = []  # Full data for all conditions.\n",
    "\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        files.append(file)\n",
    "    for file in files:\n",
    "        f = open(file)\n",
    "        data = json.load(f)\n",
    "        expr_files.append(data) \n",
    "        \n",
    "    # Extract the conditions, expression data and replicates.\n",
    "    conditions = []\n",
    "    expression = []\n",
    "    replicates = []\n",
    "    for data in expr_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        expression.append(data[\"expression\"])\n",
    "        replicates.append(data[\"replicates\"])\n",
    "        \n",
    "    # Extract all appeared genes. And because under each condition, the \n",
    "    # number of genes is(should) be the same. Only take the gene name \n",
    "    # set from the first condition.\n",
    "    genes = list(expression[0].keys())\n",
    "    \n",
    "    return conditions, expression, replicates, genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e9bad5d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_mov_data(direction):\n",
    "    \"\"\"\n",
    "    This function extracts the movement information, conditions, replicates,\n",
    "    and (GeneID, ProtID) from .JSON files containing movement results.\n",
    "    Movements are applied for calculations of final ewfd score.\n",
    "    Lists of conditions, replicates, mean_movement, and IDs will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \"\"\"\n",
    "    # Read in data:\n",
    "    os.chdir(direction)\n",
    "    files = []\n",
    "    mov_files = []\n",
    "    # Find all .json files along the direction.\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        files.append(file)\n",
    "    # Open all files and gather them together.\n",
    "    for c in files:\n",
    "        f = open(c)\n",
    "        data = json.load(f)\n",
    "        mov_files.append(data) \n",
    "    \n",
    "    # Extract condition, id and mean_mov information.\n",
    "    conditions = []\n",
    "    replicates = []\n",
    "    all_mean_mov = []\n",
    "    all_id = []\n",
    "    for data in mov_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        replicates.append(data[\"replicates\"])\n",
    "        movement = data[\"movement\"]\n",
    "        merge_id = []\n",
    "        mean_mov = []\n",
    "        for gene in movement.keys():\n",
    "            protlist = movement[gene][\"prot_ids\"]\n",
    "            mean_movlist = movement[gene][\"mean_mov\"]\n",
    "            for i in range(len(protlist)):\n",
    "                merge_id.append((gene, protlist[i]))\n",
    "                mean_mov.append(mean_movlist[i])\n",
    "        all_mean_mov.append(mean_mov)\n",
    "        all_id.append(merge_id)\n",
    "    \n",
    "    return conditions, replicates, all_mean_mov, all_id, mov_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4e8efa4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_genes(direction):\n",
    "    \"\"\"\n",
    "    This function extracts the active genes, i.e., the genes that are \n",
    "    represented in all replicates (total gene expression!= 0). \n",
    "    A list of the genes will be returned. \n",
    "    It will be recommended that applying the function for \n",
    "    once and reuse the returned list for multiple functions below.\n",
    "    The final data frame with expression values and a list of active\n",
    "    genes will be returned. \n",
    "    \n",
    "    direction: The path (string) to the folder where all gene expression \n",
    "    data are.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read the expression .json files based on the direction path. \n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "\n",
    "    DF = pd.DataFrame(genes, columns = [\"gene_id\"])\n",
    "    \n",
    "    # Extract the \"total\" gene expression over all replicates. Merge them\n",
    "    # together to form a general data frame (could be returned to view the\n",
    "    # original data). The index of the data frame are renamed replicates.\n",
    "    # The number n after the condition indicates the n-th replicate. \n",
    "    for i in range(len(conditions)):\n",
    "        expr = expression[i]\n",
    "        repl = replicates[i]\n",
    "        cond = conditions[i]\n",
    "        total_expr = np.zeros((len(genes),len(repl)))\n",
    "        for j in range(len(genes)):\n",
    "            gene = genes[j]\n",
    "            for k in range(len(repl)):\n",
    "                repl_name = repl[k]\n",
    "                total_expr[j, k] = expr[gene][repl_name][\"total\"]\n",
    "                \n",
    "        repl_names = []\n",
    "        for n in range(len(repl)):\n",
    "            repl_names.append(str(cond)+\"_\"+str(n+1))\n",
    "            \n",
    "        temp_DF = pd.DataFrame(total_expr, columns = repl_names)\n",
    "        DF = pd.concat([DF, temp_DF], axis = 1)\n",
    "        \n",
    "    # Only keep those genes that are presented in all replicates.\n",
    "    non_zero = DF[(DF.T != 0).all()]\n",
    "    active_gene = list(non_zero[\"gene_id\"])\n",
    "    \n",
    "    return non_zero, active_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "91003fb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCA_tool(DataFrame, delete_column):\n",
    "    \"\"\"\n",
    "    This function applies PCA method on the final expression DataFrame to \n",
    "    generate a DataFrame containing first three PCs for all conditions. \n",
    "    The DataFrame containing PC and information content vector will be \n",
    "    returned.\n",
    "    \n",
    "    DataFrame: The final DataFrame containing expression data for all \n",
    "    conditions.\n",
    "    delete_column: String. The name of the column in the DataFrame \n",
    "    containing geneID or (geneID, protID).\n",
    "    \n",
    "    \"\"\"\n",
    "    # Apply PCA approach on the merged data frame.\n",
    "    pca_DF = DataFrame.drop(columns = delete_column)\n",
    "    tdf = pca_DF.T\n",
    "    ids = tdf.columns\n",
    "    x = tdf.loc[:, ids].values\n",
    "    y = pd.DataFrame(tdf.index.values, columns = [\"repl\"])\n",
    "    targets = list(y[\"repl\"])\n",
    "    x = StandardScaler().fit_transform(x)\n",
    "    pca = PCA(n_components=3)\n",
    "    principalComponents = pca.fit_transform(x)\n",
    "    principalDf = pd.DataFrame(data = principalComponents\n",
    "            , columns = ['PC1', 'PC2', 'PC3']\n",
    "                               , index = y[\"repl\"])\n",
    "    # Modification of the Principal data frame to make plot easier.\n",
    "    principalDf[\"class\"] = 0\n",
    "    principalDf[\"label\"] = \"N\"\n",
    "    \n",
    "    \n",
    "    # Information content of the PCA.\n",
    "    info_content = pca.explained_variance_ratio_\n",
    "    \n",
    "    return principalDf, info_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "de18ffac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCA_plot_expression(conditions, replicates, info_content\n",
    "                        , plot_name, principalDf):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function visualize the results of PCA with the first three PCs.\n",
    "    The first Subplot is a dynamic 3D plot. Three projections with two \n",
    "    PCs will be also generated. This function could only be appliey with\n",
    "    expression data since the data structure differs from movement data(\n",
    "    applied for EWFD score). \n",
    "    \n",
    "    conditions: the list of involved conditions.\n",
    "    replicates: the list of replicates for all conditions.\n",
    "    info_content: the list of information content generated from PCA.\n",
    "    plot_name: String. The name of the plot. \n",
    "    principalDf: data frame containing PCA results. \n",
    "    \"\"\"\n",
    "    \n",
    "    # Iterativly add replicates under the same condition to the 3D plot. \n",
    "    temp_position = 0\n",
    "    fig = plt.figure()\n",
    "    fig.suptitle(plot_name, fontsize = 15)\n",
    "    \n",
    "    gs = GridSpec(18, 22)\n",
    "    \n",
    "    # First subplot.\n",
    "    ax1 = fig.add_subplot(gs[3:16,0:13], projection = \"3d\")\n",
    "    for i in range(len(conditions)):\n",
    "        repl = replicates[i]\n",
    "        for j in range(temp_position, temp_position+len(repl)):\n",
    "            principalDf[\"class\"].values[j]= i\n",
    "            principalDf[\"label\"].values[j] = conditions[i]\n",
    "\n",
    "        ax1.scatter(principalDf[\"PC1\"].values[temp_position:temp_position+len(repl)]\n",
    "                   , principalDf[\"PC2\"].values[temp_position:temp_position+len(repl)]\n",
    "                   , principalDf[\"PC3\"].values[temp_position:temp_position+len(repl)]\n",
    "                   , alpha = 1 # Turn off the transparency. \n",
    "                   , label = principalDf[\"label\"][temp_position])\n",
    "\n",
    "        temp_position += len(repl)\n",
    "    # Some general set ups to the plot. \n",
    "    ax1.set_xlabel('Principal Component 1', fontsize = 10)\n",
    "    ax1.set_ylabel('Principal Component 2', fontsize = 10)\n",
    "    ax1.set_zlabel('Principal Component 3', fontsize = 10)\n",
    "    ax1.set_title('information content:'+ str(round(sum(info_content), 3)*100)+r'%', fontsize=10)\n",
    "    ax1.legend()\n",
    "    \n",
    "    \n",
    "    # Second subplot.\n",
    "    temp_position = 0\n",
    "    ax2 = fig.add_subplot(gs[0:4,16:22]) \n",
    "    for i in range(len(conditions)):\n",
    "        repl = replicates[i]\n",
    "        ax2.scatter(principalDf[\"PC1\"].values[temp_position:temp_position+len(repl)]\n",
    "                       , principalDf[\"PC2\"].values[temp_position:temp_position+len(repl)]\n",
    "                       , label = principalDf[\"label\"][temp_position])\n",
    "        temp_position += len(repl)\n",
    "        \n",
    "    ax2.set_xlabel('Principal Component 1', fontsize = 10)\n",
    "    ax2.set_ylabel('Principal Component 2', fontsize = 10)\n",
    "    ax2.set_title('information content:'+ str(round(sum(info_content[0:2]), 2)*100)+r'%', fontsize=10)\n",
    "    ax2.legend().set_visible(False)\n",
    "    \n",
    "    # Third Subplot.\n",
    "    temp_position = 0\n",
    "    ax3 = fig.add_subplot(gs[7:11,16:22])\n",
    "    for i in range(len(conditions)):\n",
    "        repl = replicates[i]\n",
    "        ax3.scatter(principalDf[\"PC1\"].values[temp_position:temp_position+len(repl)]\n",
    "                       , principalDf[\"PC3\"].values[temp_position:temp_position+len(repl)]\n",
    "                       , label = principalDf[\"label\"][temp_position])\n",
    "        temp_position += len(repl)\n",
    "        \n",
    "    ax3.set_xlabel('Principal Component 1', fontsize = 10)\n",
    "    ax3.set_ylabel('Principal Component 3', fontsize = 10)\n",
    "    ax3.set_title('information content:'+ str(np.round(info_content[0]+info_content[2], 1)*100)+r'%', fontsize=10)\n",
    "    ax3.legend().set_visible(False)\n",
    "    \n",
    "    # Fourth Subplot.\n",
    "    temp_position = 0\n",
    "    ax4 = fig.add_subplot(gs[14:18,16:22])\n",
    "    for i in range(len(conditions)):\n",
    "        repl = replicates[i]\n",
    "        ax4.scatter(principalDf[\"PC2\"].values[temp_position:temp_position+len(repl)]\n",
    "                       , principalDf[\"PC3\"].values[temp_position:temp_position+len(repl)]\n",
    "                       , label = principalDf[\"label\"][temp_position])\n",
    "        temp_position += len(repl)\n",
    "        \n",
    "    ax4.set_xlabel('Principal Component 2', fontsize = 10)\n",
    "    ax4.set_ylabel('Principal Component 3', fontsize = 10)\n",
    "    ax4.set_title('information content:'+ str(np.round(sum(info_content[1:]), 2)*100)+r'%', fontsize=10)\n",
    "    ax4.legend().set_visible(False)\n",
    "    \n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cfc98a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCA_plot_mov(conditions, repl_num, info_content, plot_name,\n",
    "                principalDf):\n",
    "    \"\"\"\n",
    "    This function visualize the results of PCA with the first three PCs.\n",
    "    The first Subplot is a dynamic 3D plot. Three projections with two \n",
    "    PCs will be also generated. This function could only be appliey with\n",
    "    movement data (used for EWFD score calculation) since the data \n",
    "    structure differs from expression data. \n",
    "    \n",
    "    conditions: the list of involved conditions.\n",
    "    repl_num: the list of number of replicates for each condition in the \n",
    "    order conditions listed in conditions-list.\n",
    "    info_content: the list of infomation content generated from PCA.\n",
    "    plot_name: String. Name of the plot.\n",
    "    principalDf: the data frame containing PCA results\n",
    "    \"\"\"\n",
    "    \n",
    "    # Iterativly add replicates under the same condition to the 3D plot. \n",
    "    temp_position = 0\n",
    "    fig = plt.figure()\n",
    "    fig.suptitle(plot_name, fontsize = 15)\n",
    "    \n",
    "    gs = GridSpec(18, 22)\n",
    "    \n",
    "    # First Subplot\n",
    "    ax1 = fig.add_subplot(gs[3:16,0:13], projection = \"3d\")\n",
    "    for i in range(len(conditions)):\n",
    "        principalDf[\"label\"].values[i] = conditions[i]\n",
    "\n",
    "    for i in range(len(repl_num)):\n",
    "        for j in range(temp_position, temp_position+repl_num[i]):\n",
    "            principalDf[\"class\"].values[j]= i\n",
    "\n",
    "        ax1.scatter(principalDf[\"PC1\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , principalDf[\"PC2\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , principalDf[\"PC3\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , alpha = 1 # Turn off the transparency. \n",
    "                        , label = principalDf[\"label\"][temp_position])\n",
    "\n",
    "        temp_position += repl_num[i]\n",
    "    # Some general set ups to the plot. \n",
    "    ax1.set_xlabel('Principal Component 1', fontsize = 10)\n",
    "    ax1.set_ylabel('Principal Component 2', fontsize = 10)\n",
    "    ax1.set_zlabel('Principal Component 3', fontsize = 10)\n",
    "    ax1.set_title('information content:'+ str(round(sum(info_content), 2)*100)+r'%', fontsize=10)\n",
    "    ax1.legend()\n",
    "    \n",
    "    # Second Subplot\n",
    "    temp_position = 0\n",
    "    ax2 = fig.add_subplot(gs[0:4,16:22])\n",
    "    for i in range(len(repl_num)):\n",
    "        ax2.scatter(principalDf[\"PC1\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , principalDf[\"PC2\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , label = principalDf[\"label\"][temp_position])\n",
    "        temp_position += repl_num[i]\n",
    "        \n",
    "    ax2.set_xlabel('Principal Component 1', fontsize = 10)\n",
    "    ax2.set_ylabel('Principal Component 2', fontsize = 10)\n",
    "    ax2.set_title('information content:'+ str(round(sum(info_content[0:2]), 2)*100)+r'%', fontsize=10)\n",
    "    ax2.legend().set_visible(False)\n",
    "    \n",
    "    # Third Subplot\n",
    "    temp_position = 0\n",
    "    ax3 = fig.add_subplot(gs[7:11,16:22])\n",
    "    for i in range(len(repl_num)):\n",
    "        ax3.scatter(principalDf[\"PC1\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , principalDf[\"PC3\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , label = principalDf[\"label\"][temp_position])\n",
    "        temp_position += repl_num[i]\n",
    "        \n",
    "    ax3.set_xlabel('Principal Component 1', fontsize = 10)\n",
    "    ax3.set_ylabel('Principal Component 3', fontsize = 10)\n",
    "    ax3.set_title('information content:'+ str(np.round(info_content[0]+info_content[2], 2)*100)+r'%',fontsize=10)\n",
    "    ax3.legend().set_visible(False)\n",
    "    \n",
    "    \n",
    "    #Fourth Subplot\n",
    "    temp_position = 0\n",
    "    ax4 = fig.add_subplot(gs[14:18,16:22])\n",
    "    for i in range(len(repl_num)):\n",
    "        ax4.scatter(principalDf[\"PC2\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , principalDf[\"PC3\"].values[temp_position:temp_position+repl_num[i]]\n",
    "                        , label = principalDf[\"label\"][temp_position])\n",
    "        temp_position += repl_num[i]\n",
    "        \n",
    "    ax4.set_xlabel('Principal Component 2', fontsize = 10)\n",
    "    ax4.set_ylabel('Principal Component 3', fontsize = 10)\n",
    "    ax4.set_title('information content:'+ str(np.round(sum(info_content[1:]), 2)*100)+r'%', fontsize=10)\n",
    "    ax4.legend().set_visible(False)\n",
    "        \n",
    "    \n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f25ca5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_gene_expr(direction):\n",
    "    \"\"\"\n",
    "    This function visualize the results of PCA of gene expression values\n",
    "    only for active genes, i.e.,non-zero expression values in all \n",
    "    replicates.\n",
    "    A panel plot will be returned.\n",
    "    \n",
    "    direction: String. The path to the folder where all expression files \n",
    "    are stored.\n",
    "    \"\"\"\n",
    "    # Read in expression data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    \n",
    "    # Sort out the active genes and the expression DataFrame of those genes.\n",
    "    gene_set = active_genes(direction)\n",
    "    expr_DF = gene_set[0]\n",
    "    active = gene_set[1]\n",
    "    \n",
    "    # Apply the PCA on the DataFrame with only active genes.\n",
    "    results = PCA_tool(expr_DF, \"gene_id\")\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualize the results from PCA.\n",
    "    plot_name = \"Gene Expression\"\n",
    "    PCA_plot_expression(conditions, replicates, info_content\n",
    "                        , plot_name, principalDf)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "42e49941",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_transcript_expr(direction):\n",
    "    \"\"\"\n",
    "    This function visualize the results of PCA of transcript expression \n",
    "    values only for active genes, i.e.,non-zero gene expression values in \n",
    "    all replicates.\n",
    "    A panel plot will be returned.\n",
    "    \n",
    "    direction: String. The path to the folder where all expression files \n",
    "    are stored.\n",
    "    \"\"\"\n",
    "    # Read in expression data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Sort out the active genes and the expression DataFrame of those genes.\n",
    "    gene_set = active_genes(direction)\n",
    "    active = gene_set[1]\n",
    "    \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_gene = []\n",
    "    all_prot = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_gene.append(gene)\n",
    "            all_prot.append(prot)\n",
    "    \n",
    "    DF_prot = pd.DataFrame(list(zip(all_gene, all_prot))\n",
    "                           , columns = [\"gene_id\", \"prot_id\"])\n",
    "    \n",
    "    # Extract all expression values for all combinations (gene_id, prot_id)\n",
    "    # for all conditions. And build up a data frame storing all expression\n",
    "    # for each (gene_id, prot_id) under every condition.\n",
    "    for i in range(len(conditions)):\n",
    "        expr = expression[i]\n",
    "        repl = replicates[i]\n",
    "        cond = conditions[i]\n",
    "        all_prot_expr = np.zeros((len(all_prot), len(repl)))\n",
    "        for j in range(len(all_prot)):\n",
    "            gene_id = all_gene[j]\n",
    "            prot_id = all_prot[j]\n",
    "            for k in range(len(repl)):\n",
    "                repl_name = repl[k]\n",
    "                all_prot_expr[j,k] = expr[gene_id][repl_name][prot_id]\n",
    "\n",
    "        repl_names = []\n",
    "        for n in range(len(repl)):\n",
    "            repl_names.append(str(cond)+\"_\"+str(n+1))\n",
    "\n",
    "        temp_DF = pd.DataFrame(all_prot_expr, columns = repl_names)\n",
    "        DF_prot = pd.concat([DF_prot, temp_DF], axis = 1)\n",
    "        \n",
    "    # Only keep the active genes in the gene expression data.\n",
    "    transcript_DF = DF_prot[DF_prot[\"gene_id\"].isin(active)]\n",
    "    \n",
    "    # Apply the PCA.\n",
    "    results = PCA_tool(transcript_DF, [\"gene_id\",\"prot_id\"])\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualization of results from PCA.\n",
    "    plot_name = \"Transcript Expression\"\n",
    "    PCA_plot_expression(conditions, replicates, info_content\n",
    "                        , plot_name, principalDf)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fd103066",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_relative_transcript_expr(direction):\n",
    "    \"\"\"\n",
    "    The function plot the PCA of splice variant expression only for active \n",
    "    genes (i.e.the total gene expression != 0 for all replicates in gene \n",
    "    expression data). The function generate for relative splice variant \n",
    "    expression firstly a data frame which contains all isoforms. Then, \n",
    "    a sub data frame will be sorted out based on the active_gene. \n",
    "    Only those isoforms will be kept whose initial gene is active.\n",
    "    \n",
    "    direction: String. The path to the folder where all expression files \n",
    "    are stored.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read in expression data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Sort out the active genes.\n",
    "    active = active_genes(direction)[1]\n",
    "    \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_gene = []\n",
    "    all_prot = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_gene.append(gene)\n",
    "            all_prot.append(prot)\n",
    "            \n",
    "    # Build up iterativly the data frame containing all relativ splice \n",
    "    # expression under each condition. \n",
    "    DF_rel_prot = pd.DataFrame(list(zip(all_gene, all_prot)), \n",
    "                                columns = [\"gene_id\",\"prot_id\"])\n",
    "    \n",
    "    for i in range(len(conditions)):\n",
    "        expr = expression[i]\n",
    "        repl = replicates[i]\n",
    "        cond = conditions[i]\n",
    "        all_prot_expr = np.zeros((len(all_prot), len(repl)))\n",
    "        for j in range(len(all_prot)):\n",
    "            gene_id = all_gene[j]\n",
    "            prot_id = all_prot[j]\n",
    "            for k in range(len(repl)):\n",
    "                repl_name = repl[k]\n",
    "                single_gene_expr = expr[gene_id][repl_name][\"total\"]\n",
    "                if single_gene_expr != 0:\n",
    "                    all_prot_expr[j,k] = expr[gene_id][repl_name][prot_id]/single_gene_expr\n",
    "                # else: if the single gene is not expressed, then the isoforms\n",
    "                # are also not expressed. In this case, the \"score\" will be 0\n",
    "        repl_names = []\n",
    "        for n in range(len(repl)):\n",
    "            repl_names.append(str(cond)+\"_\"+str(n+1))\n",
    "\n",
    "        temp_DF = pd.DataFrame(all_prot_expr, columns = repl_names)\n",
    "        DF_rel_prot = pd.concat([DF_rel_prot, temp_DF], axis = 1)\n",
    "    \n",
    "    \n",
    "    # Only keep genes that are represented in the gene expression data.\n",
    "    rel_transcriptDF = DF_rel_prot[DF_rel_prot[\"gene_id\"].isin(active)]\n",
    "    \n",
    "    # Apply the PCA.\n",
    "    results = PCA_tool(rel_transcriptDF, [\"gene_id\",\"prot_id\"])\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualization of results from PCA.\n",
    "    plot_name = \"Relative Transcript Expression\"\n",
    "    PCA_plot_expression(conditions, replicates, info_content\n",
    "                        , plot_name, principalDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1ce6f378",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_ewfd(mov_direction, expr_direction):\n",
    "    \"\"\"\n",
    "    The function plot the PCA of EWFD scores only for active genes (i.e.\n",
    "    the total gene expression != 0 for all replicates in gene expression \n",
    "    data). The function generate for EWFD scores firstly a data frame \n",
    "    which contains all isoforms. Then, a sub data frame will be sorted \n",
    "    out based on the active_gene. Only those isoforms will be\n",
    "    kept whose initial gene is active.\n",
    "    \n",
    "    mov_direction: String. The path to the folder where all movement data \n",
    "    files are stored.\n",
    "    \n",
    "    expr_direction: String. The path to the folder where all expression \n",
    "    data files are stored. \n",
    "    \n",
    "    \"\"\"\n",
    "    # Read in movement data.\n",
    "    mov_data = read_mov_data(mov_direction)\n",
    "    conditions = mov_data[0]\n",
    "    replicates = mov_data[1]\n",
    "    all_mean_mov = mov_data[2]\n",
    "    all_id = mov_data[3]\n",
    "    mov_files = mov_data[4]\n",
    "    \n",
    "    # Sort out the active genes.\n",
    "    active = active_genes(expr_direction)[1]\n",
    "    \n",
    "    # Extract condition, id and mean_mov information.\n",
    "    conditions = []\n",
    "    replicates = []\n",
    "    all_mean_mov = []\n",
    "    all_gene_id = []\n",
    "    all_prot_id = []\n",
    "    for data in mov_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        replicates.append(data[\"replicates\"])\n",
    "        movement = data[\"movement\"]\n",
    "        merge_gene_id = []\n",
    "        merge_prot_id = []\n",
    "        mean_mov = []\n",
    "        for gene in movement.keys():\n",
    "            protlist = movement[gene][\"prot_ids\"]\n",
    "            mean_movlist = movement[gene][\"mean_mov\"]\n",
    "            for i in range(len(protlist)):\n",
    "                merge_gene_id.append(gene)\n",
    "                merge_prot_id.append(protlist[i])\n",
    "                mean_mov.append(mean_movlist[i])\n",
    "        all_mean_mov.append(mean_mov)\n",
    "        all_gene_id.append(merge_gene_id)\n",
    "        all_prot_id.append(merge_prot_id)\n",
    "        \n",
    "    # Determine the number of replicates for each condition.\n",
    "    repl_num_dict = {i:conditions.count(i) for i in conditions}\n",
    "    repl_num = list(repl_num_dict.values())\n",
    "        \n",
    "    # Construct the dataframe containing all information.\n",
    "    # First with two columns:\n",
    "    EWFD_table = pd.DataFrame(list(zip(all_gene_id[0], all_prot_id[0],\\\n",
    "                                       1-np.asarray(all_mean_mov[0]))),\\\n",
    "                        columns = [\"gene_id\", \"prot_id\"\n",
    "                                   , (conditions[0], replicates[0])])\n",
    "    # Iterativly add other condition-mean_EWFD columns:\n",
    "    for i in range(1,len(conditions)):\n",
    "        EWFD_table[(conditions[i], replicates[i])] = 1-np.asarray(all_mean_mov[i])\n",
    "        \n",
    "    # Only keep represented/active genes from gene expression data. \n",
    "    active_gene_ewfd = EWFD_table[EWFD_table[\"gene_id\"].isin(active)]\n",
    "    \n",
    "    # Apply PCA.\n",
    "    results = PCA_tool(active_gene_ewfd, [\"gene_id\",\"prot_id\"])\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualization of results from PCA.\n",
    "    plot_name = \"EWFD Score\"\n",
    "    PCA_plot_mov(conditions, repl_num, info_content, plot_name,\n",
    "                principalDf)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
