{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1c1952d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob,os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "%matplotlib widget\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import json\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from mpl_toolkits.mplot3d import Axes3D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d74fe57e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_expression_data(direction):\n",
    "    \"\"\"\n",
    "    This function read in all .JSON files along the given direction. \n",
    "    Involved conditions, expression data under each condition,\n",
    "    replicates under each condition, and the set of genes will be extracted. \n",
    "    They will be returned as lists.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read the expression .json files based on the direction path. \n",
    "    os.chdir(direction)\n",
    "    files = []\n",
    "    expr_files = []  # Full data for all conditions.\n",
    "\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        files.append(file)\n",
    "    for file in files:\n",
    "        f = open(file)\n",
    "        data = json.load(f)\n",
    "        expr_files.append(data) \n",
    "        \n",
    "    # Extract the conditions, expression data and replicates.\n",
    "    conditions = []\n",
    "    expression = []\n",
    "    replicates = []\n",
    "    for data in expr_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        expression.append(data[\"expression\"])\n",
    "        replicates.append(data[\"replicates\"])\n",
    "        \n",
    "    # Extract all appeared genes. And because under each condition, the \n",
    "    # number of genes is(should) be the same. Only take the gene name \n",
    "    # set from the first condition.\n",
    "    genes = list(expression[0].keys())\n",
    "    \n",
    "    return conditions, expression, replicates, genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ead15b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_mov_data(direction):\n",
    "    \"\"\"\n",
    "    This function extracts the movement information, conditions, replicates,\n",
    "    and (GeneID, ProtID) from .JSON files containing movement results.\n",
    "    Movements are applied for calculations of final ewfd score.\n",
    "    Lists of conditions, replicates, mean_movement, and IDs will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \"\"\"\n",
    "    # Read in all_movement data \n",
    "    os.chdir(direction)\n",
    "    cond = []\n",
    "    mov_files = []\n",
    "    # Find all .json files along the direction.\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        cond.append(file)\n",
    "    # Open all files and gather them together.\n",
    "    for c in cond:\n",
    "        f = open(c)\n",
    "        data = json.load(f)\n",
    "        mov_files.append(data)\n",
    "        \n",
    "        \n",
    "    # Extract condition, id and mean_mov information.\n",
    "    conditions = []\n",
    "    all_mean_mov = [] \n",
    "    all_gene_id = []\n",
    "    all_prot_id = []\n",
    "    for data in mov_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        movement = data[\"movement\"]\n",
    "        gene_id = []\n",
    "        prot_id = []\n",
    "        mean_mov = []\n",
    "        for gene in movement.keys():\n",
    "            protlist = movement[gene][\"prot_ids\"]\n",
    "            mean_movlist = movement[gene][\"mean_mov\"]\n",
    "            for i in range(len(protlist)):\n",
    "                gene_id.append(gene)\n",
    "                prot_id.append(protlist[i])\n",
    "                mean_mov.append(mean_movlist[i])\n",
    "        all_mean_mov.append(mean_mov)\n",
    "        all_gene_id.append(gene_id)\n",
    "        all_prot_id.append(prot_id)\n",
    "        \n",
    "    return all_mean_mov, all_gene_id, all_prot_id, conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2210b20a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_genes(direction):\n",
    "    \"\"\"\n",
    "    This function extracts the active genes, i.e., the genes that are \n",
    "    represented in all replicates (total gene expression!= 0). \n",
    "    A list of the genes will be returned. \n",
    "    It will be recommended that applying the function for \n",
    "    once and reuse the returned list for multiple functions below.\n",
    "    The final data frame with expression values and a list of active\n",
    "    genes will be returned. \n",
    "    \n",
    "    direction: The path (string) to the folder where all gene expression \n",
    "    data are.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read the expression .json files based on the direction path. \n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "\n",
    "    DF = pd.DataFrame(genes, columns = [\"gene_id\"])\n",
    "    \n",
    "    # Extract the \"total\" gene expression over all replicates. Merge them\n",
    "    # together to form a general data frame (could be returned to view the\n",
    "    # original data). The index of the data frame are renamed replicates.\n",
    "    # The number n after the condition indicates the n-th replicate. \n",
    "    for i in range(len(conditions)):\n",
    "        expr = expression[i]\n",
    "        repl = replicates[i]\n",
    "        cond = conditions[i]\n",
    "        total_expr = np.zeros((len(genes),len(repl)))\n",
    "        for j in range(len(genes)):\n",
    "            gene = genes[j]\n",
    "            for k in range(len(repl)):\n",
    "                repl_name = repl[k]\n",
    "                total_expr[j, k] = expr[gene][repl_name][\"total\"]\n",
    "                \n",
    "        repl_names = []\n",
    "        for n in range(len(repl)):\n",
    "            repl_names.append(str(cond)+\"_\"+str(n+1))\n",
    "            \n",
    "        temp_DF = pd.DataFrame(total_expr, columns = repl_names)\n",
    "        DF = pd.concat([DF, temp_DF], axis = 1)\n",
    "        \n",
    "    # Only keep those genes that are presented in all replicates.\n",
    "    non_zero = DF[(DF.T != 0).all()]\n",
    "    active_gene = list(non_zero[\"gene_id\"])\n",
    "    \n",
    "    return non_zero, active_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3dc6e822",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCA_tool(DataFrame, delete_column):\n",
    "    \"\"\"\n",
    "    This function applies PCA method on the final expression DataFrame to \n",
    "    generate a DataFrame containing first three PCs for all conditions. \n",
    "    The DataFrame containing PC and information content vector will be \n",
    "    returned.\n",
    "    \n",
    "    DataFrame: The final DataFrame containing expression data for all \n",
    "    conditions.\n",
    "    delete_column: String. The name of the column in the DataFrame \n",
    "    containing geneID or (geneID, protID).\n",
    "    \n",
    "    \"\"\"\n",
    "        \n",
    "    # Apply the 3-dim PCA.\n",
    "    # Coding PCA with the help of:\n",
    "    # https://towardsdatascience.com/pca-using-python-scikit-learn-e653f8989e60\n",
    "    pca_DF = DataFrame.drop(columns = delete_column)\n",
    "    tdf = pca_DF.T\n",
    "    ids = tdf.columns\n",
    "    x = tdf.loc[:, ids].values\n",
    "    y = pd.DataFrame(tdf.index.values, columns = [\"conditions\"])\n",
    "    targets = list(y[\"conditions\"])\n",
    "    x = StandardScaler().fit_transform(x) # Turn x into normalized scale.\n",
    "    pca = PCA(n_components=3)\n",
    "    principalComponents = pca.fit_transform(x)\n",
    "    principalDf = pd.DataFrame(data = principalComponents\n",
    "                 , columns = ['PC1', \\\n",
    "                              'PC2', \\\n",
    "                              'PC3'])\n",
    "    # FinalDF contains the 3 components and the conditions as columns.\n",
    "    finalDf = pd.concat([principalDf, y], axis = 1)  \n",
    "    # The information contain of 3 components.\n",
    "    info_contain = pca.explained_variance_ratio_\n",
    "    \n",
    "    return finalDf, info_contain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ea94ef7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCA_plot(conditions, info_content, plot_name, principalDf):\n",
    "    \"\"\"\n",
    "    This function visualize the results of PCA with the first three PCs.     The first Subplot is a dynamic 3D plot. Three projections with two \n",
    "    PCs will be also generated. This function could only be appliey with\n",
    "    expression data since the data structure differs from movement data(\n",
    "    applied for EWFD score). \n",
    "    \n",
    "    conditions: the list of involved conditions.\n",
    "    replicates: the list of replicates for all conditions.\n",
    "    info_content: the list of information content generated from PCA.\n",
    "    plot_name: String. The name of the plot. \n",
    "    principalDf: data frame containing PCA results. \n",
    "    \n",
    "    \"\"\"\n",
    "        \n",
    "    # Construct the plots:\n",
    "\n",
    "    fig = plt.figure()\n",
    "    fig.suptitle(plot_name, fontsize = 15)\n",
    "\n",
    "    gs = GridSpec(18, 22)\n",
    "    \n",
    "    x = principalDf[\"PC1\"]\n",
    "    y = principalDf[\"PC2\"]\n",
    "    z = principalDf[\"PC3\"]\n",
    "    name = principalDf[\"conditions\"]\n",
    "\n",
    "    # First subplot.\n",
    "    ax1 = fig.add_subplot(gs[3:16,0:13], projection='3d')\n",
    "    ax1.set_xlabel('PC1', fontsize = 10)\n",
    "    ax1.set_ylabel('PC2', fontsize = 10)\n",
    "    ax1.set_zlabel('PC3', fontsize = 10)\n",
    "    ax1.set_title('information content:'+\\\n",
    "                     str(np.round(sum(info_content), 2)*100)+\"%\"\n",
    "                     ,fontsize=10)\n",
    "    ax1.scatter(x, y ,z)\n",
    "    for i, txt in enumerate(name):\n",
    "        ax1.text(x[i], y[i], z[i], '%s'%txt) \n",
    "\n",
    "\n",
    "    # Second subplot.\n",
    "    ax2 = fig.add_subplot(gs[0:4,16:22])\n",
    "    ax2.set_xlabel('PC1', fontsize = 8)\n",
    "    ax2.set_ylabel('PC2', fontsize = 8)\n",
    "    ax2.scatter(x, y)\n",
    "    ax2.set_title('information content:'+\\\n",
    "                    str(np.round(sum(info_content[0:2]), 2)*100)+\"%\", \\\n",
    "                  fontsize=8)\n",
    "    for i, txt in enumerate(name):\n",
    "        ax2.text(x[i], y[i], '%s'%txt, size=8) \n",
    "\n",
    "    # Third subplot.\n",
    "    ax3 = fig.add_subplot(gs[7:11,16:22])\n",
    "    ax3.set_xlabel('PC1', fontsize = 8)\n",
    "    ax3.set_ylabel('PC3', fontsize = 8)\n",
    "    ax3.scatter(x, z)\n",
    "    ax3.set_title('information content:'+\\\n",
    "                    str(np.round(info_content[0]+info_content[2], 2)*100)+\"%\"\n",
    "                  , fontsize=8)\n",
    "    for i, txt in enumerate(name):\n",
    "        ax3.text(x[i], z[i], '%s'%txt, size=8) \n",
    "\n",
    "    # Forth subplot.\n",
    "    ax4 = fig.add_subplot(gs[14:18,16:22])\n",
    "    ax4.set_xlabel('PC2', fontsize = 8)\n",
    "    ax4.set_ylabel('PC3', fontsize = 8)\n",
    "    ax4.scatter(y, z)\n",
    "    ax4.set_title('information content:'+\\\n",
    "                    str(np.round(sum(info_content[1:]), 2)*100)+\"%\"\n",
    "                  , fontsize=8)\n",
    "    for i, txt in enumerate(name):\n",
    "        ax4.text(y[i], z[i], '%s'%txt, size=8) \n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "7ecef07e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_gene_mean(direction):\n",
    "    \"\"\"\n",
    "    This function generates firstly a DataFrame containing all mean values\n",
    "    for gene expression and keep those only for active genes. \n",
    "    The mean values for each condition are calculated through averaging over\n",
    "    all replicates.\n",
    "    PCA is applied for dimensional reduction and allows visualization.\n",
    "    \n",
    "    direction: String. Path to the folder where all expression files are \n",
    "    stored.\n",
    "    \"\"\"\n",
    "    # Read in expression data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Extract active genes\n",
    "    gene_set = active_genes(direction)\n",
    "    active = gene_set[1]\n",
    "    \n",
    "    # Calculate the mean(gene expression i.e. \"total\") for each gene over\n",
    "    # all replicates. Do this for each condition. The dataframe could be \n",
    "    # returned to virsulize the data/vectors as table. \n",
    "    all_mean_total_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        cond = expression[i]\n",
    "        repl = replicates[i]\n",
    "        total_expr = np.zeros(len(genes))\n",
    "        for j in range(len(genes)):\n",
    "            gene = genes[j]\n",
    "            repl_expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                repl_name = repl[k]\n",
    "                repl_expr[k] = cond[gene][repl_name][\"total\"]\n",
    "            total_expr[j] = np.mean(repl_expr)\n",
    "        all_mean_total_expr.append(total_expr)\n",
    "    # The dataframe to present mean gene expression for all conditions.\n",
    "    all_gene_expr_DF = pd.DataFrame(genes, columns = [\"gene_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        all_gene_expr_DF[conditions[i]] = all_mean_total_expr[i]\n",
    "        \n",
    "    # Only keep genes that are represented in the gene expression data.\n",
    "    expr_DF = all_gene_expr_DF[all_gene_expr_DF[\"gene_id\"].isin(active)]\n",
    "    \n",
    "    # Apply PCA on expression DataFrame\n",
    "    results = PCA_tool(expr_DF, \"gene_id\")\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualization of PCA results.\n",
    "    plot_name = \"Gene Expression\"\n",
    "    PCA_plot(conditions, info_content, plot_name, principalDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5582d99b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_transcript_mean(direction):\n",
    "    \"\"\"\n",
    "    The function plots the PCA of mean splice variant expreesion (transcript \n",
    "    expression) for each condition but only for active genes, i.e., those\n",
    "    are present in gene expression by all replicates.\n",
    "    Mean values are calculated through averaging over all replicates.\n",
    "    \n",
    "    dierction: the path (string) to the folder where all expression data\n",
    "    are.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read in expression data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Extract active genes\n",
    "    gene_set = active_genes(direction)\n",
    "    active = gene_set[1]\n",
    "    \n",
    "        \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_gene = []\n",
    "    all_prot = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_gene.append(gene)\n",
    "            all_prot.append(prot)\n",
    "    \n",
    "    \n",
    "    # Extract all expression values for all combinations (gene_id, prot_id)\n",
    "    # for all conditions.\n",
    "    total_prot_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        test = expression[i]\n",
    "        repl = replicates[i]\n",
    "        all_prot_expr = np.zeros(len(all_prot))\n",
    "        for j in range(len(all_prot)):\n",
    "            gene_id = all_gene[j]\n",
    "            prot_id = all_prot[j]\n",
    "            expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                expr[k] = test[gene_id][repl[k]][prot_id]\n",
    "            all_prot_expr[j] = np.mean(expr)\n",
    "        total_prot_expr.append(all_prot_expr)\n",
    "    # Construct the data frame from the data/vectors above. \n",
    "    # The dataframe could be returned to visualize the data.\n",
    "    prot_expr_DF = pd.DataFrame(list(zip(all_gene, all_prot)), \\\n",
    "                                columns = [\"gene_id\", \"prot_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        prot_expr_DF[conditions[i]] = total_prot_expr[i]\n",
    "        \n",
    "    # Only keep genes that are active in gene expression data (in all replicates)\n",
    "    transcript_DF = prot_expr_DF[prot_expr_DF[\"gene_id\"].isin(active)]\n",
    "    \n",
    "    # Apply PCA on the final expression DataFrame.\n",
    "    results = PCA_tool(transcript_DF, [\"gene_id\", \"prot_id\"])\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualization of results from PCA.\n",
    "    plot_name = \"Transcript Expression\"\n",
    "    PCA_plot(conditions, info_content, plot_name, principalDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c5cfe2e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_rel_transcript_mean(direction):\n",
    "    \"\"\"\n",
    "    The function plots PCA results of mean relative splice variant \n",
    "    expression (transcript expression). The mean is calculated with:\n",
    "    (sum_{replicates} isoform expression/gene expression) / #replicates.\n",
    "    This is also the values that are applied for calculating EWFD values.\n",
    "    Only active genes (from gene expression) will be considered in \n",
    "    the analysis.\n",
    "    \n",
    "    direction:  the path (string) to the folder where all expression data\n",
    "    are.\n",
    "    \"\"\"\n",
    "    # Read in expression data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Extract active genes\n",
    "    gene_set = active_genes(direction)\n",
    "    active = gene_set[1]\n",
    "    \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_gene = []\n",
    "    all_prot = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_gene.append(gene)\n",
    "            all_prot.append(prot)\n",
    "    \n",
    "    # Calculate the vectors with the mean of fractions \n",
    "    # sum(isof/gene)/#replicates\n",
    "    total_rel_prot_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        test = expression[i]\n",
    "        repl = replicates[i]\n",
    "        all_prot_expr = np.zeros(len(all_prot))\n",
    "        for j in range(len(all_prot)):\n",
    "            gene_id = all_gene[j]\n",
    "            prot_id = all_prot[j]\n",
    "            expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                single_gene_expr = test[gene_id][repl[k]][\"total\"]\n",
    "                if single_gene_expr != 0:\n",
    "                    expr[k] = test[gene_id][repl[k]][prot_id]/single_gene_expr\n",
    "                # else: if the single gene is not expressed, then the isoforms\n",
    "                # are also not expressed. In this case, the \"score\" will be 0.\n",
    "            all_prot_expr[j] = np.mean(expr)\n",
    "        total_rel_prot_expr.append(all_prot_expr)\n",
    "    # Construct the dataframe from the calculation.\n",
    "    rel_prot_expr_DF = pd.DataFrame(list(zip(all_gene, all_prot)), \\\n",
    "                                    columns = [\"gene_id\", \"prot_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        rel_prot_expr_DF[conditions[i]] = total_rel_prot_expr[i]\n",
    "        \n",
    "    # Only keep active genes from gene expression data (in all replicates).\n",
    "    rel_transcript_DF = rel_prot_expr_DF[rel_prot_expr_DF[\"gene_id\"].isin(active)]\n",
    "    \n",
    "    # Apply PCA.\n",
    "    results = PCA_tool(rel_transcript_DF, [\"gene_id\", \"prot_id\"])\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualization of results from PCA.\n",
    "    plot_name = \"Relative Transcript Expression\"\n",
    "    PCA_plot(conditions, info_content, plot_name, principalDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e4eeafa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def active_ewfd_mean(mov_direction, expr_direction):\n",
    "    \"\"\"\n",
    "    This function applies mean values of ewfd scores calculated from\n",
    "    movement values. Only active genes are involved. Visualizing the \n",
    "    data through dimensional reduction with PCA returned as a panel image.\n",
    "    \n",
    "    mov_direction: String. The path to the folder where all movement data \n",
    "    files are stored.\n",
    "    \n",
    "    expr_direction: String. The path to the folder where all expression data \n",
    "    files are stored.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read in movement data.\n",
    "    mov_results = read_mov_data(mov_direction)\n",
    "    \n",
    "    all_mean_mov = mov_results[0]\n",
    "    all_gene_id = mov_results[1]\n",
    "    all_prot_id = mov_results[2]\n",
    "    conditions = mov_results[3]\n",
    "    \n",
    "    # Extract active genes.\n",
    "    gene_set = active_genes(expr_direction)\n",
    "    active = gene_set[1]\n",
    "    \n",
    "    \n",
    "    # Construct the dataframe containing all information.\n",
    "    # First with two columns:\n",
    "    EWFD_table = pd.DataFrame(list(zip(all_gene_id[0]\n",
    "                                       , all_prot_id[0]\n",
    "                                       , 1-np.asarray(all_mean_mov[0])))\n",
    "                            ,columns = [\"gene_id\", \"prot_id\"\n",
    "                                        , conditions[0]])\n",
    "        \n",
    "    # Iterativly add other condition-mean_mov columns:\n",
    "    for i in range(1,len(conditions)):\n",
    "        EWFD_table[conditions[i]] = 1-np.asarray(all_mean_mov[i])\n",
    "    \n",
    "    # Only keep genes that are represented in the gene expression data.\n",
    "    ewfd_DF = EWFD_table[EWFD_table[\"gene_id\"].isin(active)]\n",
    "    \n",
    "    \n",
    "    # Apply PCA.\n",
    "    results = PCA_tool(ewfd_DF, [\"gene_id\", \"prot_id\"])\n",
    "    principalDf = results[0]\n",
    "    info_content = results[1]\n",
    "    \n",
    "    # Visualization of results from PCA.\n",
    "    plot_name = \"EWFD Score\"\n",
    "    PCA_plot(conditions, info_content, plot_name, principalDf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
