{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "65ceebe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob,os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "%matplotlib widget\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import json\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.cluster import KMeans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "75d9cbdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_expression_data(direction):\n",
    "    \"\"\"\n",
    "    This function read in all .JSON files along the given direction. \n",
    "    Involved conditions, expression data under each condition,\n",
    "    replicates under each condition, and the set of genes will be extracted. \n",
    "    They will be returned as lists.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read the expression .json files based on the direction path.\n",
    "    os.chdir(direction)\n",
    "    files = []\n",
    "    expr_files = []  # Full data for all conditions.\n",
    "\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        files.append(file)\n",
    "\n",
    "    for file in files:\n",
    "        f = open(file)\n",
    "        data = json.load(f)\n",
    "        expr_files.append(data)  \n",
    "        \n",
    "        \n",
    "    # Extract the conditions, expression data and replicates.\n",
    "    conditions = []\n",
    "    expression = []\n",
    "    replicates = []\n",
    "    for data in expr_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        expression.append(data[\"expression\"])\n",
    "        replicates.append(data[\"replicates\"])\n",
    "        \n",
    "        \n",
    "    # Extract all appeared genes. And because under each condition, the \n",
    "    # number of genes is(should) be the same. Only take the gene name \n",
    "    # set from the first condition.\n",
    "    genes = list(expression[0].keys())\n",
    "    \n",
    "    return conditions, expression, replicates, genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "824d0753",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_mov_data(direction):\n",
    "    \"\"\"\n",
    "    This function extracts the movement information, conditions, replicates,\n",
    "    and (GeneID, ProtID) from .JSON files containing movement results.\n",
    "    Movements are applied for calculations of final ewfd score.\n",
    "    Lists of conditions, mean_movement, and IDs will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \"\"\"\n",
    "    # Read in all_movement data \n",
    "    os.chdir(direction)\n",
    "    cond = []\n",
    "    mov_files = []\n",
    "    # Find all .json files along the direction.\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        cond.append(file)\n",
    "    # Open all files and gather them together.\n",
    "    for c in cond:\n",
    "        f = open(c)\n",
    "        data = json.load(f)\n",
    "        mov_files.append(data)\n",
    "        \n",
    "        \n",
    "    # Extract condition, id and mean_mov information.\n",
    "    conditions = []\n",
    "    all_mean_mov = []\n",
    "    all_id = []\n",
    "    for data in mov_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        movement = data[\"movement\"]\n",
    "        merge_id = []\n",
    "        mean_mov = []\n",
    "        for gene in movement.keys():\n",
    "            protlist = movement[gene][\"prot_ids\"]\n",
    "            mean_movlist = movement[gene][\"mean_mov\"]\n",
    "            for i in range(len(protlist)):\n",
    "                merge_id.append((gene, protlist[i]))\n",
    "                mean_mov.append(mean_movlist[i])\n",
    "        all_mean_mov.append(mean_mov)\n",
    "        all_id.append(merge_id)\n",
    "        \n",
    "    return conditions, all_mean_mov, all_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98802a30",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCA_kmeans_tool(DataFrame, delete_column, n_cluster):\n",
    "    \"\"\"\n",
    "    This function applies PCA method on the final expression DataFrame to \n",
    "    generate a DataFrame containing first three PCs for all conditions. \n",
    "    The DataFrame containing PC and information content vector will be \n",
    "    returned.\n",
    "    The results from PCA are then treated with k-means clustering algorithm.\n",
    "    The number of clusters is given (should normaly equal to the number of \n",
    "    cell types or number of conditions).\n",
    "    \n",
    "    DataFrame: The final DataFrame containing expression data for all \n",
    "    conditions.\n",
    "    delete_column: String. The name of the column in the DataFrame \n",
    "    containing geneID or (geneID, protID).\n",
    "    n_cluster: The number of clusters that the data points should be \n",
    "    clustered by k-means.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Apply the 3-dim PCA.\n",
    "    # Coding PCA with the help of:\n",
    "    # https://towardsdatascience.com/pca-using-python-scikit-learn-e653f8989e60\n",
    "    tdf = DataFrame.T\n",
    "    col_index = tdf.loc[delete_column,:]\n",
    "    df2 = tdf.rename(columns = col_index)\n",
    "    df3 = df2.drop(delete_column)\n",
    "    ids = df3.columns\n",
    "    x = df3.loc[:, ids].values\n",
    "    y = pd.DataFrame(df3.index.values, columns = [\"conditions\"])\n",
    "    targets = list(y[\"conditions\"])\n",
    "    x = StandardScaler().fit_transform(x) # Turn x into normalized scale.\n",
    "    pca = PCA(n_components=3)\n",
    "    principalComponents = pca.fit_transform(x)\n",
    "    principalDf = pd.DataFrame(data = principalComponents\n",
    "                 , columns = ['principal component 1', \\\n",
    "                              'principal component 2', \\\n",
    "                              'principal component 3'])\n",
    "    # FinalDF contains the 3 components and the conditions as columns.\n",
    "    finalDf = pd.concat([principalDf, y], axis = 1)  \n",
    "    # The information contain of 3 components.\n",
    "    info_contain = pca.explained_variance_ratio_\n",
    "    \n",
    "    \n",
    "    # Apply the k-Mean approach to clustering.\n",
    "    # Coding k-Mean with the help of:\n",
    "    # https://jakevdp.github.io/PythonDataScienceHandbook/05.11-k-means.html\n",
    "    kmeanDf = np.array(list(zip(principalDf[\"principal component 1\"], \\\n",
    "                 principalDf[\"principal component 2\"],\\\n",
    "                 principalDf[\"principal component 3\"])))\n",
    "    kmeans = KMeans(n_clusters=n_cluster).fit(kmeanDf)\n",
    "    y_kmeans = kmeans.predict(kmeanDf)\n",
    "    name = finalDf[\"conditions\"]\n",
    "\n",
    "    \n",
    "    return kmeanDf, info_contain, y_kmeans, name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6c24b70d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def results_plot(kmeanDf, info_content, plot_name, y_kmeans, name):\n",
    "    \"\"\"\n",
    "    This function visualize the results of PCA after treating k-means \n",
    "    clustering algorithm. \n",
    "    Provided the k-means results data frame, the function returns\n",
    "    a panel image with one dynamic 3D plot and three 2D projections.\n",
    "    \n",
    "    kmeanDf: the result DataFrame applied by k-means. Also the \n",
    "    DataFrame generated by PCA.\n",
    "    info_content: the list of information content generated from PCA.\n",
    "    plot_name: String. The name of the plot. \n",
    "    y_kmeans: the results of kmean.predict() function. \n",
    "    name: list/array of condition names in the DataFrame applied in PCA.\n",
    "\n",
    "    \"\"\"\n",
    "    x = kmeanDf[:, 0]\n",
    "    y = kmeanDf[:, 1]\n",
    "    z = kmeanDf[:, 2]\n",
    "    \n",
    "    # Contruct the plots:\n",
    "    fig = plt.figure()\n",
    "    fig.suptitle(plot_name, fontsize = 15)\n",
    "\n",
    "    gs = GridSpec(18, 22)\n",
    "\n",
    "\n",
    "    # First subplot.\n",
    "    ax1 = fig.add_subplot(gs[3:16,0:13], projection='3d')\n",
    "    ax1.set_xlabel('Principal Component 1', fontsize = 10)\n",
    "    ax1.set_ylabel('Principal Component 2', fontsize = 10)\n",
    "    ax1.set_zlabel('Principal Component 3', fontsize = 10)\n",
    "    ax1.set_title('information contain:'+\\\n",
    "                     str(np.round(sum(info_content), 2)*100)+\\\n",
    "                     r'%   number of clusters:'+\\\n",
    "                     str(2), \\\n",
    "                     fontsize=10)\n",
    "    ax1.scatter(x, y, z, c=y_kmeans, s=50, cmap='viridis')\n",
    "    for i, txt in enumerate(name):\n",
    "        ax1.text(x[i], y[i], z[i], '%s'%txt) \n",
    "\n",
    "\n",
    "    # Second subplot.\n",
    "    ax2 = fig.add_subplot(gs[0:4,16:22])\n",
    "    ax2.set_xlabel('Principal Component 1', fontsize = 8)\n",
    "    ax2.set_ylabel('Principal Component 2', fontsize = 8)\n",
    "    ax2.scatter(x, y, c=y_kmeans, s=50, cmap='viridis')\n",
    "    ax2.set_title('information contain:'+\\\n",
    "                    str(np.round(sum(info_content[0:2]), 2)*100)+\"%\", \\\n",
    "                  fontsize=8)\n",
    "    for i, txt in enumerate(name):\n",
    "        ax2.text(x[i], y[i], '%s'%txt, size=8) \n",
    "\n",
    "    # Third subplot.\n",
    "    ax3 = fig.add_subplot(gs[7:11,16:22])\n",
    "    ax3.set_xlabel('Principal Component 1', fontsize = 8)\n",
    "    ax3.set_ylabel('Principal Component 3', fontsize = 8)\n",
    "    ax3.scatter(x, z, c=y_kmeans, s=50, cmap='viridis')\n",
    "    ax3.set_title('information contain:'+\\\n",
    "                    str(np.round(info_content[0]+info_content[2], 2)*100)+\"%\", \\\n",
    "                  fontsize=8)\n",
    "    for i, txt in enumerate(name):\n",
    "        ax3.text(x[i], z[i], '%s'%txt, size=8) \n",
    "\n",
    "    # Forth subplot.\n",
    "    ax4 = fig.add_subplot(gs[14:18,16:22])\n",
    "    ax4.set_xlabel('Principal Component 2', fontsize = 8)\n",
    "    ax4.set_ylabel('Principal Component 3', fontsize = 8)\n",
    "    ax4.scatter(y, z, c=y_kmeans, s=50, cmap='viridis')\n",
    "    ax4.set_title('information contain:'+\\\n",
    "                    str(np.round(sum(info_content[1:]), 2)*100)+\"%\", fontsize=8)\n",
    "    for i, txt in enumerate(name):\n",
    "        ax4.text(y[i], z[i], '%s'%txt, size=8) \n",
    "\n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2b7e4494",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_expr_mean(direction):\n",
    "    \"\"\"\n",
    "    This function will cluster the conditions based on the original gene \n",
    "    expression. Here, the mean gene expression value over replicates\n",
    "    for each gene will be applied to construct vectors for each condition. \n",
    "    After that, the PCA approach will project the data to 3 dimensional \n",
    "    space and k-Mean will be used to cluster the conditions. \n",
    "    The output is a 3-dim interactive plot with n_cluster colored clusters.\n",
    "    \n",
    "    direction: the path(str) to the folder where all .json expression \n",
    "    files are involved.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Calculate the mean(gene expression i.e. \"total\") for each gene over\n",
    "    # all replicates. Do this for each condition. The dataframe could be \n",
    "    # returned to virsulize the data/vectors as table. \n",
    "    all_mean_total_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        cond = expression[i]\n",
    "        repl = replicates[i]\n",
    "        total_expr = np.zeros(len(genes))\n",
    "        for j in range(len(genes)):\n",
    "            gene = genes[j]\n",
    "            repl_expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                repl_name = repl[k]\n",
    "                repl_expr[k] = cond[gene][repl_name][\"total\"]\n",
    "            total_expr[j] = np.mean(repl_expr)\n",
    "        all_mean_total_expr.append(total_expr)\n",
    "    # The dataframe to present mean gene expression for all conditions.\n",
    "    all_gene_expr_DF = pd.DataFrame(genes, columns = [\"gene_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        all_gene_expr_DF[conditions[i]] = all_mean_total_expr[i]\n",
    "        \n",
    "    # Apply PCA and k-means algorithm.\n",
    "    # Set number of clusters equals 2.\n",
    "    results = PCA_kmeans_tool(all_gene_expr_DF, \"gene_id\", 2)\n",
    "    kmeanDf = results[0]\n",
    "    info_content = results[1]\n",
    "    y_kmeans = results[2]\n",
    "    name = results[3]\n",
    "    \n",
    "    # Plot the results from PCA and k-means.\n",
    "    plot_name = \"Gene Expression\"\n",
    "    results_plot(kmeanDf, info_content, plot_name, y_kmeans, name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a1fd5829",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcript_expr_mean(direction):\n",
    "    \"\"\"\n",
    "    This function read all expression .json files along the given direction.\n",
    "    Based on the protein expression vectors for each condition, the PCA \n",
    "    approach project the data to 3 dimensional space. k-Mean will be applied\n",
    "    to cluster the conditions.\n",
    "    The output will be a 3-dim plot with the first 3 PCA components as \n",
    "    coordinations and colored clusters.\n",
    "    \n",
    "    direction: the path(str) to the folder where all .json expression\n",
    "    files are involved.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_combine = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_combine.append((gene, prot))\n",
    "    \n",
    "    \n",
    "    \n",
    "    # Extract all expression values for all combinations (gene_id, prot_id)\n",
    "    # for all conditions.\n",
    "    total_prot_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        test = expression[i]\n",
    "        repl = replicates[i]\n",
    "        all_prot_expr = np.zeros(len(all_combine))\n",
    "        for j in range(len(all_combine)):\n",
    "            comb = all_combine[j]\n",
    "            gene_id = comb[0]\n",
    "            prot_id = comb[1]\n",
    "            expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                expr[k] = test[gene_id][repl[k]][prot_id]\n",
    "            all_prot_expr[j] = np.mean(expr)\n",
    "        total_prot_expr.append(all_prot_expr)\n",
    "    # Construct the data frame from the data/vectors above. \n",
    "    # The dataframe could be returned to visualize the data.\n",
    "    prot_expr_DF = pd.DataFrame(list(zip(all_combine)), \\\n",
    "                                columns = [\"gene_id, prot_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        prot_expr_DF[conditions[i]] = total_prot_expr[i]\n",
    "        \n",
    "    \n",
    "    # Apply PCA and k-means clustering.\n",
    "    # Set number of clusters equals 2.\n",
    "    results = PCA_kmeans_tool(prot_expr_DF, \"gene_id, prot_id\", 2)\n",
    "    kmeanDf = results[0]\n",
    "    info_content = results[1]\n",
    "    y_kmeans = results[2]\n",
    "    name = results[3]\n",
    "    \n",
    "    # Plot the results from PCA and k-means.\n",
    "    plot_name = \"Transcript Expression\"\n",
    "    results_plot(kmeanDf, info_content, plot_name, y_kmeans, name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "62661ad1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def relative_transcript_expr(direction):\n",
    "    \"\"\"\n",
    "    This function clusters the conditions based on the relative protein \n",
    "    expression: sum(isoform/gene)/#replicates for 1 isoform. \n",
    "    PCA approach will be applied to project the data to 3 dimensional space.\n",
    "    k-Mean will be used to cluster the conditions. \n",
    "    The output will be a 3 dimensional plot with colored clusters.\n",
    "    \n",
    "    direction: the path (str) to all expression .json files. \n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_combine = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_combine.append((gene, prot))\n",
    "            \n",
    "    \n",
    "    # Calculate the vectors with the mean of fractions \n",
    "    # sum(isof/gene)/#replicates\n",
    "    total_rel_prot_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        test = expression[i]\n",
    "        repl = replicates[i]\n",
    "        all_prot_expr = np.zeros(len(all_combine))\n",
    "        for j in range(len(all_combine)):\n",
    "            comb = all_combine[j]\n",
    "            gene_id = comb[0]\n",
    "            prot_id = comb[1]\n",
    "            expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                single_gene_expr = test[gene_id][repl[k]][\"total\"]\n",
    "                if single_gene_expr != 0:\n",
    "                    expr[k] = test[gene_id][repl[k]][prot_id]/single_gene_expr\n",
    "                # else: if the single gene is not expressed, then the isoforms\n",
    "                # are also not expressed. In this case, the \"score\" will be 0.\n",
    "            all_prot_expr[j] = np.mean(expr)\n",
    "        total_rel_prot_expr.append(all_prot_expr)\n",
    "    # Construct the dataframe from the calculation.\n",
    "    rel_prot_expr_DF = pd.DataFrame(list(zip(all_combine)), \\\n",
    "                                    columns = [\"gene_id, prot_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        rel_prot_expr_DF[conditions[i]] = total_rel_prot_expr[i]\n",
    "    \n",
    "    \n",
    "    # Apply PCA and k-means clustering.\n",
    "    # Set number of clusters equals 2.\n",
    "    results = PCA_kmeans_tool(rel_prot_expr_DF, \"gene_id, prot_id\", 2)\n",
    "    kmeanDf = results[0]\n",
    "    info_content = results[1]\n",
    "    y_kmeans = results[2]\n",
    "    name = results[3]\n",
    "    \n",
    "    # Plot the results from PCA and k-means.\n",
    "    plot_name = \"Relative Transcript Expression\"\n",
    "    results_plot(kmeanDf, info_content, plot_name, y_kmeans, name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0e753094",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ewfd_mean(direction):\n",
    "    \"\"\"\n",
    "    The function firstly read all .json files from the direction.\n",
    "    EWFD values will be calculated and the DataFrame will be generated.\n",
    "    After the application of pca approach on the dataset, the conditions \n",
    "    will be virsulized on a 2-dim space. The k-Mean algorithm will be used \n",
    "    to cluster the conditions.\n",
    "    \n",
    "    direction: the string of the path to .json movment files. \n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    mov_data = read_mov_data(direction)\n",
    "    \n",
    "    conditions = mov_data[0]\n",
    "    all_mean_mov = mov_data[1]\n",
    "    all_id = mov_data[2]\n",
    "    \n",
    "    # Construct the dataframe containing all information.\n",
    "    # First with two columns:\n",
    "    EWFD_table = pd.DataFrame(list(zip(all_id[0], 1-np.asarray(all_mean_mov[0]))),\\\n",
    "                            columns = [\"gene_id, prot_id\", conditions[0]])\n",
    "        \n",
    "    # Iterativly add other condition-mean_mov columns:\n",
    "    for i in range(1,len(conditions)):\n",
    "        EWFD_table[conditions[i]] = 1-np.asarray(all_mean_mov[i])\n",
    "        \n",
    "    # Apply PCA and k-means clustering.\n",
    "    # Set number of clusters equals 2.\n",
    "    results = PCA_kmeans_tool(EWFD_table, \"gene_id, prot_id\", 2)\n",
    "    kmeanDf = results[0]\n",
    "    info_content = results[1]\n",
    "    y_kmeans = results[2]\n",
    "    name = results[3]\n",
    "    \n",
    "    # Plot the results from PCA and k-means.\n",
    "    plot_name = \"Relative Transcript Expression\"\n",
    "    results_plot(kmeanDf, info_content, plot_name, y_kmeans, name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
