{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "07c226ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob,os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "get_ipython().run_line_magic('matplotlib', 'widget')\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import json\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.cluster import KMeans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a92222bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_expression_data(direction):\n",
    "    \"\"\"\n",
    "    This function read in all .JSON files along the given direction. \n",
    "    Involved conditions, expression data under each condition,\n",
    "    replicates under each condition, and the set of genes will be extracted. \n",
    "    They will be returned as lists.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read the expression .json files based on the direction path.\n",
    "    os.chdir(direction)\n",
    "    files = []\n",
    "    expr_files = []  # Full data for all conditions.\n",
    "\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        files.append(file)\n",
    "\n",
    "    for file in files:\n",
    "        f = open(file)\n",
    "        data = json.load(f)\n",
    "        expr_files.append(data)  \n",
    "        \n",
    "        \n",
    "    # Extract the conditions, expression data and replicates.\n",
    "    conditions = []\n",
    "    expression = []\n",
    "    replicates = []\n",
    "    for data in expr_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        expression.append(data[\"expression\"])\n",
    "        replicates.append(data[\"replicates\"])\n",
    "        \n",
    "        \n",
    "    # Extract all appeared genes. And because under each condition, the \n",
    "    # number of genes is(should) be the same. Only take the gene name \n",
    "    # set from the first condition.\n",
    "    genes = list(expression[0].keys())\n",
    "    \n",
    "    return conditions, expression, replicates, genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d1c2675c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_mov_data(direction):\n",
    "    \"\"\"\n",
    "    This function extracts the movement information, conditions, replicates,\n",
    "    and (GeneID, ProtID) from .JSON files containing movement results.\n",
    "    Movements are applied for calculations of final ewfd score.\n",
    "    Lists of conditions, mean_movement, and IDs will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder where all .JSON files are stored.\n",
    "    \"\"\"\n",
    "    # Read in all_movement data \n",
    "    os.chdir(direction)\n",
    "    cond = []\n",
    "    mov_files = []\n",
    "    # Find all .json files along the direction.\n",
    "    for file in glob.glob(\"*.json\"):\n",
    "        cond.append(file)\n",
    "    # Open all files and gather them together.\n",
    "    for c in cond:\n",
    "        f = open(c)\n",
    "        data = json.load(f)\n",
    "        mov_files.append(data)\n",
    "        \n",
    "        \n",
    "    # Extract condition, id and mean_mov information.\n",
    "    conditions = []\n",
    "    all_mean_mov = []\n",
    "    all_id = []\n",
    "    for data in mov_files:\n",
    "        conditions.append(data[\"condition\"])\n",
    "        movement = data[\"movement\"]\n",
    "        merge_id = []\n",
    "        mean_mov = []\n",
    "        for gene in movement.keys():\n",
    "            protlist = movement[gene][\"prot_ids\"]\n",
    "            mean_movlist = movement[gene][\"mean_mov\"]\n",
    "            for i in range(len(protlist)):\n",
    "                merge_id.append((gene, protlist[i]))\n",
    "                mean_mov.append(mean_movlist[i])\n",
    "        all_mean_mov.append(mean_mov)\n",
    "        all_id.append(merge_id)\n",
    "        \n",
    "    return conditions, all_mean_mov, all_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1cbcad33",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PCA_tool(DataFrame, delete_column, conditions):\n",
    "    \"\"\"\n",
    "    This function applies PCA on the given DataFrame and calculate all \n",
    "    the principal components for each condition. The over all information\n",
    "    content is also returned.\n",
    "    Results are returned as dictionary format. \n",
    "    \n",
    "    DataFrame: the pandas DataFrame applied with PCA.\n",
    "    delete_column: the index for the column with GeneID or (GeneID, ProtID).\n",
    "    conditions: the list of conditions. \n",
    "    \n",
    "    \"\"\"\n",
    "    # Apply the 3-dim PCA.\n",
    "    # Coding PCA with the help of:\n",
    "    # https://towardsdatascience.com/pca-using-python-scikit-learn-e653f8989e60\n",
    "    tdf = DataFrame.T\n",
    "    col_index = tdf.loc[delete_column,:]\n",
    "    df2 = tdf.rename(columns = col_index)\n",
    "    df3 = df2.drop(delete_column)\n",
    "    ids = df3.columns\n",
    "    x = df3.loc[:, ids].values\n",
    "    y = pd.DataFrame(df3.index.values, columns = [\"conditions\"])\n",
    "    targets = list(y[\"conditions\"])\n",
    "    x = StandardScaler().fit_transform(x) # Turn x into normalized scale.\n",
    "    \n",
    "    n_conditions = len(conditions)\n",
    "    pca = PCA(n_components=n_conditions)\n",
    "    column = []\n",
    "    for i in range(n_conditions):\n",
    "        column.append(\"PC\"+str(i+1))\n",
    "    principalComponents = pca.fit_transform(x)\n",
    "    principalDf = pd.DataFrame(data = principalComponents\n",
    "                 , columns = column, index = conditions)\n",
    "    # Variables coordinates of each condition. \n",
    "    result_principal = principalDf.to_dict(\"index\")\n",
    "    # Vector for information contains.\n",
    "    info_content = pca.explained_variance_ratio_\n",
    "    info_contentDf = pd.DataFrame(data = info_content\n",
    "                              , columns = [\"information_content\"]\n",
    "                              , index = column).T\n",
    "    result_info = info_contentDf.to_dict(\"index\")[\"information_content\"]\n",
    "    \n",
    "    return result_info, result_principal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2b3887aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PC_gene_expr(direction):\n",
    "    \"\"\"\n",
    "    This function generates all principal components for all conditions\n",
    "    from PCA of gene expression.\n",
    "    Also the overall information content vector is included.\n",
    "    A dictionary of the information will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder of all expression files.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Calculate the mean(gene expression i.e. \"total\") for each gene over\n",
    "    # all replicates. Do this for each condition. The dataframe could be \n",
    "    # returned to virsulize the data/vectors as table. \n",
    "    all_mean_total_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        cond = expression[i]\n",
    "        repl = replicates[i]\n",
    "        total_expr = np.zeros(len(genes))\n",
    "        for j in range(len(genes)):\n",
    "            gene = genes[j]\n",
    "            repl_expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                repl_name = repl[k]\n",
    "                repl_expr[k] = cond[gene][repl_name][\"total\"]\n",
    "            total_expr[j] = np.mean(repl_expr)\n",
    "        all_mean_total_expr.append(total_expr)\n",
    "    # The dataframe to present mean gene expression for all conditions.\n",
    "    all_gene_expr_DF = pd.DataFrame(genes, columns = [\"gene_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        all_gene_expr_DF[conditions[i]] = all_mean_total_expr[i]\n",
    "        \n",
    "    # Apply PCA.\n",
    "    results = PCA_tool(all_gene_expr_DF, \"gene_id\", conditions)\n",
    "    result_info = results[0]\n",
    "    result_principal = results[1]\n",
    "    \n",
    "    return {\"information_content\":result_info, \"conditions\":result_principal}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0456150a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PC_transcript_expr(direction):\n",
    "    \"\"\"\n",
    "    This function generates all principal components for all conditions\n",
    "    from PCA of transcript expression.\n",
    "    Also the overall information content vector is included.\n",
    "    A dictionary of the information will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder of all expression files.\n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_combine = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_combine.append((gene, prot))\n",
    "    \n",
    "    \n",
    "    # Extract all expression values for all combinations (gene_id, prot_id)\n",
    "    # for all conditions.\n",
    "    total_prot_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        test = expression[i]\n",
    "        repl = replicates[i]\n",
    "        all_prot_expr = np.zeros(len(all_combine))\n",
    "        for j in range(len(all_combine)):\n",
    "            comb = all_combine[j]\n",
    "            gene_id = comb[0]\n",
    "            prot_id = comb[1]\n",
    "            expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                expr[k] = test[gene_id][repl[k]][prot_id]\n",
    "            all_prot_expr[j] = np.mean(expr)\n",
    "        total_prot_expr.append(all_prot_expr)\n",
    "    # Construct the data frame from the data/vectors above. \n",
    "    # The dataframe could be returned to visualize the data.\n",
    "    prot_expr_DF = pd.DataFrame(list(zip(all_combine)), \\\n",
    "                                columns = [\"gene_id, prot_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        prot_expr_DF[conditions[i]] = total_prot_expr[i]\n",
    "        \n",
    "        \n",
    "    # Apply PCA.\n",
    "    results = PCA_tool(prot_expr_DF, \"gene_id, prot_id\", conditions)\n",
    "    result_info = results[0]\n",
    "    result_principal = results[1]\n",
    "    \n",
    "    return {\"information_content\":result_info, \"conditions\":result_principal}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f849096e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PC_rel_transcript_expr(direction):\n",
    "    \"\"\"\n",
    "    This function generates all principal components for all conditions\n",
    "    from PCA of relative transcript expression.\n",
    "    Also the overall information content vector is included.\n",
    "    A dictionary of the information will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder of all expression files. \n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    expr_data = read_expression_data(direction)\n",
    "    \n",
    "    conditions = expr_data[0]\n",
    "    expression = expr_data[1]\n",
    "    replicates = expr_data[2]\n",
    "    genes = expr_data[3]\n",
    "    \n",
    "    # Extract all combinations (gene_id, prot_id). Because all .json files\n",
    "    # have(should have) the same number of combinations. Could only take\n",
    "    # the set of all combinations from data under the first condition. \n",
    "    all_combine = []\n",
    "    data_cond1 = expression[0]\n",
    "    repl = replicates[0][0]  # Replicates of the first condition.\n",
    "    for i in range(len(genes)):\n",
    "        gene = genes[i]\n",
    "        keys = list(data_cond1[gene][repl].keys())\n",
    "        for prot in keys[1:]:  # In every key list, the first element is \"total\"\n",
    "            all_combine.append((gene, prot))\n",
    "            \n",
    "    \n",
    "    # Calculate the vectors with the mean of fractions \n",
    "    # sum(isof/gene)/#replicates\n",
    "    total_rel_prot_expr = []\n",
    "    for i in range(len(conditions)):\n",
    "        test = expression[i]\n",
    "        repl = replicates[i]\n",
    "        all_prot_expr = np.zeros(len(all_combine))\n",
    "        for j in range(len(all_combine)):\n",
    "            comb = all_combine[j]\n",
    "            gene_id = comb[0]\n",
    "            prot_id = comb[1]\n",
    "            expr = np.zeros(len(repl))\n",
    "            for k in range(len(repl)):\n",
    "                single_gene_expr = test[gene_id][repl[k]][\"total\"]\n",
    "                if single_gene_expr != 0:\n",
    "                    expr[k] = test[gene_id][repl[k]][prot_id]/single_gene_expr\n",
    "                # else: if the single gene is not expressed, then the isoforms\n",
    "                # are also not expressed. In this case, the \"score\" will be 0.\n",
    "            all_prot_expr[j] = np.mean(expr)\n",
    "        total_rel_prot_expr.append(all_prot_expr)\n",
    "    # Construct the dataframe from the calculation.\n",
    "    rel_prot_expr_DF = pd.DataFrame(list(zip(all_combine)), \\\n",
    "                                    columns = [\"gene_id, prot_id\"])\n",
    "    for i in range(len(conditions)):\n",
    "        rel_prot_expr_DF[conditions[i]] = total_rel_prot_expr[i]\n",
    "    \n",
    "    # Apply PCA.\n",
    "    results = PCA_tool(rel_prot_expr_DF, \"gene_id, prot_id\", conditions)\n",
    "    result_info = results[0]\n",
    "    result_principal = results[1]\n",
    "    \n",
    "    return {\"information_content\":result_info, \"conditions\":result_principal}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "76fe4228",
   "metadata": {},
   "outputs": [],
   "source": [
    "def PC_EWFD(direction):\n",
    "    \"\"\"\n",
    "    This function generates all principal components for all conditions\n",
    "    from PCA of EWFD scores.\n",
    "    Also the overall information content vector is included.\n",
    "    A dictionary of the information will be returned.\n",
    "    \n",
    "    direction: String. Path to the folder of all movement files. \n",
    "    \"\"\"\n",
    "    # Read in data.\n",
    "    mov_data = read_mov_data(direction)\n",
    "    \n",
    "    conditions = mov_data[0]\n",
    "    all_mean_mov = mov_data[1]\n",
    "    all_id = mov_data[2]\n",
    "    \n",
    "    # Check if all files have the same id-orders and number of combinations.\n",
    "    for i in range(len(conditions)-1):\n",
    "        if all_id[i] != all_id[i+1]:\n",
    "            print(\"The number of combination or id-order is not the same!\")\n",
    "    # If id id lists are the same, then just take the first one.\n",
    "    \n",
    "    \n",
    "    # Construct the dataframe containing all information.\n",
    "    # First with two columns:\n",
    "    EWFD_table = pd.DataFrame(list(zip(all_id[0], 1-np.asarray(all_mean_mov[0]))), \\\n",
    "                            columns = [\"gene_id, prot_id\", conditions[0]])\n",
    "    # Iterativly add other condition-mean_mov columns:\n",
    "    for i in range(1,len(conditions)):\n",
    "        EWFD_table[conditions[i]] = 1-np.asarray(all_mean_mov[i])\n",
    "    \n",
    "\n",
    "    # Apply PCA.\n",
    "    results = PCA_tool(EWFD_table, \"gene_id, prot_id\", conditions)\n",
    "    result_info = results[0]\n",
    "    result_principal = results[1]\n",
    "    \n",
    "    return {\"information_content\":result_info, \"conditions\":result_principal}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "af722041",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_JSON(expr_direction, mov_direction, writepath):\n",
    "    \"\"\"\n",
    "    This function generate a .JSON file containing all principal components\n",
    "    and information content for all four analysis levels.\n",
    "    \n",
    "    expr_direction: String. The path to the folder with all expression files.\n",
    "    mov_direction: String. The path to the folder with all movement files.\n",
    "    writepath: String. The path where the .JSON file should be generated. It\n",
    "    has the format of \"folder/name.json\".\n",
    "    \"\"\"\n",
    "    Data = {\"gene_expr_PCA\":PC_gene_expr(expr_direction),\\\n",
    "            \"transcript_expr_PCA\":PC_transcript_expr(expr_direction),\\\n",
    "            \"relative_transcript_expr_PCA\":PC_rel_transcript_expr(expr_direction),\\\n",
    "            \"EWDF_PCA\": PC_EWFD(mov_direction)}\n",
    "    pca_pc = json.dumps(Data, indent=4)  \n",
    "    mode = 'a' if os.path.exists(writepath) else 'w'\n",
    "    with open(writepath, mode) as f:\n",
    "        f.write(pca_pc)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
